contracts/command-batcher/src/contract.rs

```diff
@@ -1,44 +1,343 @@
+        let config = CONFIG.load(deps.storage)?;
+
+        let query = gateway::msg::QueryMsg::GetMessages { message_ids };
+        let messages: Vec<Message> = deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: config.gateway.into(),
+            msg: to_binary(&query)?,
+        }))?;
+
+        if messages.is_empty() {
+            return Err(ContractError::NoMessagesFound {});
+        }
+
+        let message_ids: Vec<String> = messages.iter().map(|msg| msg.id.clone()).collect();
+        let batch_id = BatchID::new(&message_ids);
+
+        let command_batch = match COMMANDS_BATCH.may_load(deps.storage, &batch_id)? {
+            Some(batch) => batch,
+            None => {
+                let batch = CommandBatch::new(messages, config.destination_chain_id)?;
+
+                COMMANDS_BATCH.save(deps.storage, &batch.id, &batch)?;
+
+                batch
+            }
+        };
+
+        let reply_id = REPLY_ID_COUNTER.update(deps.storage, |mut reply_id| -> StdResult<_> {
+            reply_id += 1;
+            Ok(reply_id)
+        })?;
+        REPLY_ID_TO_BATCH.save(deps.storage, reply_id, &command_batch.id)?;
+
+        let start_sig_msg = multisig::msg::ExecuteMsg::StartSigningSession {
+            key_id: config.service_name,
+            msg: command_batch.msg_to_sign(),
+        };
+
+        let wasm_msg = wasm_execute(config.multisig, &start_sig_msg, vec![])?;
+
+        Ok(Response::new().add_submessage(SubMsg::reply_on_success(wasm_msg, reply_id)))
+    }
+
+    pub fn rotate_snapshot(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        pub_keys: HashMap<String, HexBinary>,
+    ) -> Result<Response, ContractError> {
+        let config = CONFIG.load(deps.storage)?;
+        if config.admin != info.sender {
+            return Err(ContractError::Unauthorized {});
+        }
+
+        let snapshot = snapshot(deps.querier, env.block, &config)?;
+
+        for participant in snapshot.participants.keys() {
+            if !pub_keys.contains_key(participant) {
+                return Err(ContractError::PublicKeyNotFound {
+                    participant: participant.to_owned(),
+                });
+            }
+        }
+
+        let keygen_msg = multisig::msg::ExecuteMsg::KeyGen {
+            key_id: config.service_name,
+            snapshot,
+            pub_keys,
+        };
+
+        Ok(Response::new().add_message(WasmMsg::Execute {
+            contract_addr: config.multisig.into(),
+            msg: to_binary(&keygen_msg)?,
+            funds: vec![],
+        }))
+    }
+
+    fn snapshot(
+        querier: QuerierWrapper,
+        block: BlockInfo,
+        config: &Config,
+    ) -> Result<Snapshot, ContractError> {
+        let query_msg = service_registry::msg::QueryMsg::GetActiveWorkers {
+            service_name: config.service_name.clone(),
+            chain_name: config.chain_name.to_string(),
```

Comment:
We need a way to distinguish the workers who are signing from the workers who are voting. Right now, they are the same, since both components use the same service name and chain name. I think they should keep using the same service, so that way their stake can be reused. But the chain name has to be different in some way. I sort of think we should append "-signing" to the chain name here (and maybe do the same for voting), so workers can signal whether they support signing and/or voting.

Reply:
Actually, I think it's fine to just have the same validator set for voting and signing for a particular chain. Makes the security easier to reason about, since there's one set of workers securing the connection, regardless of which way messages are flowing.